CONTINUATION.md

```diff
@@ -0,0 +1,511 @@
+
+```js
+let reject
+let p1 // (1)
+ctx.run('init', () => {
+  ;({ p1, reject } = Promise.withResolvers())
+})
+
+ctx.run('reject', () => {
+  reject('error message')
+})
+
+addEventListener('unhandledrejection', (event) => {
+  ctx.get() // (2)
+})
+```
+
+In this case, the `unhandledrejection` event will be dispatched with the promise
+instance `p1` and a string of `'error message'`, and the event is scheduled in
+the context of `'reject'`.
+
+For the two type of context variables with `unhandledrejection` event of `p1`:
+
+- For `AsyncContext.Variable`, the context is `'reject'`, where `p1` was rejected.
+- For `ContinuationFlow`, the context is `'reject'`, where `p1` was rejected.
+
+---
+
+However, if this promise was handled, and the new promise didn't have a proper
+handler:
+
+```js
+let reject
+let p1 // (1)
+ctx.run('init', () => {
+  ;({ p1, reject } = Promise.withResolvers())
+  const p2 = p1 // p1 is not settled yet
+    .then(undefined, undefined)
+})
+
+ctx.run('reject', () => {
+  reject('error message')
+})
+
+addEventListener('unhandledrejection', (event) => {
+  ctx.get() // (2)
+})
+```
+
+The `unhandledrejection` event will be dispatched with the promise
+instance `p2` with a string of `'error message'`. In this case, the event is
+scheduled in a new microtask which was scheduled in the context of `'reject'`.
+
+For the two type of context variables, the value at site (2) with
+`unhandledrejection` event of `p2`:
+
+- For `AsyncContext.Variable`, the context is `'init'`, where `p2`
+  attaches the promise handlers to `p1`.
+- For `ContinuationFlow`, the context is `'reject'`, where `p1` was rejected.
+
+---
+
+If handlers were attached to an already rejected promise:
+
+```js
+let p1 // (1)
+ctx.run('reject', () => {
+  p1 = Promise.reject('error message')
+})
+
+ctx.run('init', () => {
+  const p2 = p1 // p1 is already rejected
+    .then(undefined, undefined) // (2)
+})
+
+addEventListener('unhandledrejection', (event) => {
+  ctx.get() // (2)
+})
+```
+
+In this case, the `unhandledrejection` event is scheduled in the context where
+`.then` is called, that is `'init'`.
+
+> By saying "the `unhandledrejection` event is scheduled" above, it is
+> referring to [HostPromiseRejectionTracker](https://html.spec.whatwg.org/#the-hostpromiserejectiontracker-implementation).
+> This host hook didn't actually "schedule" the event dispatching, rather putting the
+> promise in a queue and the event is actually scheduled from [event loop](https://html.spec.whatwg.org/#notify-about-rejected-promises).
+
+For the two type of context variables, the value at site (2) with
+`unhandledrejection` event of `p2`:
+
+- For `AsyncContext.Variable`, the context is `'init'`, where `p2`
+  attaches the promise handlers to `p1`.
+- For `ContinuationFlow`, the context is `'reject'`, where `p1` was rejected.
+
+### Fulfilled promises
+
+Similar to the rejected promise issue, the MOST relevant cause's context when
```

Comment:
I think we should abandon the idea of "most relevant cause", since that leads to problematic heuristics.  Instead, the goal should be that the winning context is clearly defined, preferably with as simple/straightforward a ruleset as possible.  I think this is one point in `AsyncContext.Variable`'s favor, since "the context is snapshotted the moment any function to schedule is passed to a builtin" is very clearly specified with no ambiguity, whereas any other context is going to be much more situational.

Reply:
In this section, it is showing the example where the most relevant cause is not observable from JavaScript at the moment and why it could be undeterministic. I rephrased the section to highlight the intention.

Reply:
"the context is snapshotted the moment any function to schedule is passed to a builtin" is not strictly correct for what _exists_ when considering async/await, which is exactly what this is all about. A snapshot is captured when an await expression is made. It has nothing to do with any function being passed to something as you could await _anything_.

That aside though, capture and restore points are consistent with through flow semantics, and restore points are actually the same between the two. The only difference is that snapshot captures at the _end_ of a branch rather than the _start_. In the case of a singular branch and merge, the execution is linear so the expected behaviour is fairly obvious that the context flows forward through the branch and out the other end. In the case of multiple branches merging such as with `Promise.all(...)` there are some additional complexities, but a winner can be picked with no special handling in that the resolution of the aggregate promise would occur as a direct result of the final inner branch completion and so would just adopt that context automatically unless we did something else to _change_ that behaviour.

We _could_ have merge contexts to express multiple context origins, but it would still be entirely possible to "pick a winner" to at least be the _default_ value, which seems like a reasonable trade-off. The matter of if we should even _have_ a merge context can be a separate issue, and is certainly lower-priority than the base flow-through semantics.